### ISSUE TITLE
filters after training 

### ISSUE THREAD

**Issue by yyx1107** [he-y/filter-pruning-geometric-median#9]
m.do_similar_mask()
net = m.model 
After these operations('do_similar_mask'),this model's  several filters have been set to zeros.
This paper is based on soft filters pruning, but after training, I found the 'model.paramters' didn't update.What is the reasons?Or Where did I get it wrong? Thanks~
![Screenshot from 2019-06-21 16-02-26](https://user-images.githubusercontent.com/50829614/59907659-2fa90500-943e-11e9-997c-3596b6e6f9c7.png)

![Screenshot from 2019-06-21 15-57-51](https://user-images.githubusercontent.com/50829614/59907386-8bbf5980-943d-11e9-9dd7-f326d9937132.png)

**Comment by he-y** [comment]
Your observation is right. The operation of the gradient of FPGM is different from that of [SFP](https://github.com/he-y/soft-filter-pruning).
For FPGM, the gradients of the pruned filters are zero, and for SFP, they are non-zero, refer to this [link](https://github.com/he-y/filter-pruning-geometric-median/blob/master/pruning_imagenet.py#L660-L669).

Several reasons for this:

1. The nonzero gradient of SFP would enlarge the model capacity, which is the main point of SFP. But the contribution of FPGM is about the filter selection criterion, so we would like the eliminate the effect of model capacity to compare fairly. 

2. This kind of operation is also different from HFP (Please refer to HFP in my IJCAI paper). Although the selected filters are pruned and kept as zero, the connections of them remain. 

Imagine the first case, we use 500 neurons to learn a training set for one epoch (This is HFP). 
In the second case, we use 1000 neurons to learn and delete 500 of them. For every epoch, 500 neurons are different (This is SFP).
The third case, we use 1000 neurons to learn and delete 500 of them. For every epoch, 500 neurons are the same (This is something between SFP and HFP).

**Comment by yyx1107** [comment]
Got it! Thank you~

**Comment by xingyueye** [comment]
@he-y What if only change the filter selection method and make the selected filters different for every epoch?In other words, maintaining the advantages of both the model capability enlarge of SFP and the filter selection of FPGM?

**Comment by he-y** [comment]
> @he-y What if only change the filter selection method and make the selected filters different for every epoch?In other words, maintaining the advantages of both the model capability enlarge of SFP and the filter selection of FPGM?

I think the performance would be slightly better. 
As discussed above, the original FPGM is "FPGM + half SFP". So the improvement of "FPGM + SFP" over "FPGM + half SFP" may be less than that of "SFP" over "HFP".

**Comment by lizhenstat** [comment]
@he-y Hi,  Is the HFP you referred is the following paper?
Meta Filter Pruning to Accelerate Deep Convolutional Neural Networks
